src/Utils/BlockTemplateUtils.php

```diff
@@ -281,4 +281,55 @@ public static function supports_block_templates() {
 
 		return true;
 	}
+
+	/**
+	 * Checks if we can fallback to the `archive-product` template for a given slug
+	 *
+	 * `taxonomy-product_cat` and `taxonomy-product_tag` templates can generally use the
+	 * `archive-product` as a fallback if there are no specific overrides.
+	 *
+	 * @param string $template_slug Slug to check for fallbacks.
+	 * @return boolean
+	 */
+	public static function template_is_eligible_for_product_archive_fallback( $template_slug ) {
+		$eligible_for_fallbacks = array( 'taxonomy-product_cat', 'taxonomy-product_tag' );
+
+		return in_array( $template_slug, $eligible_for_fallbacks, true )
+			&& ! self::theme_has_template( $template_slug )
+			&& self::theme_has_template( 'archive-product' );
+	}
+
+	/**
+	 * Sets the `has_theme_file` to `true` for templates with fallbacks
+	 *
+	 * There are cases (such as tags and categories) in which fallback templates
+	 * can be used; so, while *technically* the theme doesn't have a specific file
+	 * for them, it is important that we tell Gutenberg that we do, in fact,
+	 * have a theme file (i.e. the fallback one).
+	 *
+	 * **Note:** this function changes the array that has been passed.
+	 *
+	 * It returns `true` if anything was changed, `false` otherwise.
+	 *
+	 * @param array $query_result Array of template objects.
+	 * @param array $template A specific template object which could have a fallback.
+	 *
+	 * @return boolean
+	 */
+	public static function set_has_theme_file_if_fallback_is_available( $query_result, $template ) {
+		foreach ( $query_result as $i => &$query_result_template ) {
```

Comment:
Looks much better but `$i` is not used

```suggestion
		foreach ( $query_result as $query_result_template ) {
```

Reply:
🤦‍♀️  I had used the index, but then I realized I didn't need that. I think I still need to pass the query_result_template as reference because I edit it, right? Again, my lackluster PHP knowledge her, but I thought the `&` was required here.

Reply:
Oh right, yeah you do sorry I missed that! I didn't notice the `&` there.

Reply:
At least that's what I understand from [the docs](https://www.php.net/manual/en/control-structures.foreach.php):

> In order to be able to directly modify array elements within the loop precede `$value` with `&`